src/lib/methods/batch/batch-write.ts

```diff
@@ -0,0 +1,102 @@
+import { Executable } from '../executable';
+import { DynamoDB } from '../../dynamodb';
+import { BatchItem } from './batch-item';
+import { BatchMethod } from './batch-method';
+import { BatchWriteItemInput, BatchWriteItemOutput } from 'aws-sdk/clients/dynamodb';
+
+export class BatchWrite extends BatchMethod implements Executable {
+
+	constructor(
+		dynamodb: DynamoDB,
+		private items: BatchItem[],
+		private msDelay = 1,
+		private retryCount = 10) {
+		super(dynamodb);
+	}
+	/**
+	 * Initialize the `InsertItem` object.
+	 */
+	initialize(): BatchWrite {
+		const request = {
+			RequestItems: {}
+		};
+		this.items.forEach((item) => {
+			const table = request.RequestItems[item.table];
+			if (!table) {
+				request.RequestItems[item.table] = [item.value];
+			} else {
+				(request.RequestItems[item.table] as any[]).push(item.value);
+			}
+		});
+		this.params = request;
+		return this;
+	}
+	/**
+	 * Builds and returns the raw DynamoDB query object.
+	 */
+	buildRawQuery() {
+		this.initialize();
+		return this.params;
+	}
+	/**
+	 * Execute the batch write request.
+	 */
+	exec(): Promise<any> {
+		const db = this.dynamodb.dynamodb;
+
+		if (!db) {
+			return Promise.reject(new Error('Call .connect() before executing queries.'));
+		}
+
+		if (this.items.length < 1) {
+			return Promise.reject(new Error('Items array can not be empty.'));
+		}
+
+		if (this.items.length > 25) {
+			return Promise.reject(new Error('Can not insert more than 25 items at a time.'));
+		}
+
+		if (!this.items) {
+			return Promise.reject(new Error('params object was undefined.'));
+		}
+
+		this.buildRawQuery();
+		const query = this.params as BatchWriteItemInput;
+
+		return this.runQuery(() => db.batchWrite(query).promise())
```

Comment:
I'm wondering if this entire block can't be rewritten with [p-retry](https://github.com/sindresorhus/p-retry#readme) which has built-in exponential back-off and works properly with the provided retry options.

Reply:
With @GentileFulvio help we managed to refactor this to use pRetry.